# 0.1 Setup

# Q. 1 - Create a new notebook for this task. Be sure to import the relevant libraries, along with your ords_prods_merge dataframe, which should include your newly derived columns from the previous Exercise.

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Define path to project folders

path = r'/Users/joefeuille/CF Analysis/Instacart Basket Analysis'

In [3]:
# Setting exported df from Task 4.7 as file as ords_prods_merge

ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ord_prod_mrg_derived.pkl'))

# Q.2 - For the entire dataframe, find the aggregated mean of the “order_number” column grouped by “department_id” for the df.

### Checking df

In [4]:
ords_prods_merge.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0  2539329.0      1.0    prior           1.0                 2.0   
1  2398795.0      1.0    prior           2.0                 3.0   
2   473747.0      1.0    prior           3.0                 3.0   
3  2254736.0      1.0    prior           4.0                 4.0   
4   431534.0      1.0    prior           5.0                 4.0   

   order_time_of_day  days_since_prior_order first_time_ordering  product_id  \
0                8.0                     NaN                True         196   
1                7.0                    15.0               False         196   
2               12.0                    21.0               False         196   
3                7.0                    29.0               False         196   
4               15.0                    28.0               False         196   

   add_to_cart_order  reordered product_name  aisle_id  department_id  prices  \
0                1.0        0.0         Soda      77.0            7.0     9.0   
1                1.0        1.0         Soda      77.0            7.0     9.0   
2                1.0        1.0         Soda      77.0            7.0     9.0   
3                1.0        1.0         Soda      77.0            7.0     9.0   
4                1.0        1.0         Soda      77.0            7.0     9.0   

  _merge    price_range_loc     busiest_day    busiest_days  \
0   both  Mid-range product  Regularly busy  Regularly busy   
1   both  Mid-range product  Regularly busy    Slowest days   
2   both  Mid-range product  Regularly busy    Slowest days   
3   both  Mid-range product      Least busy    Slowest days   
4   both  Mid-range product      Least busy    Slowest days   

  busiest_period_of_day  
0        Average orders  
1        Average orders  
2           Most orders  
3        Average orders  
4           Most orders

In [5]:
ords_prods_merge.shape

(32435070, 20)

In [6]:
# Grouping df by department_id, and using agg to find the mean for order_number


ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1.0              15.457838
2.0              17.277920
3.0              17.170395
4.0              17.811403
5.0              15.215751
6.0              16.439806
7.0              17.225802
8.0              15.340650
9.0              15.895474
10.0             20.197148
11.0             16.170638
12.0             15.887671
13.0             16.583536
14.0             16.773669
15.0             16.165037
16.0             17.665606
17.0             15.694469
18.0             19.310397
19.0             17.177343
20.0             16.473447
21.0             22.902379

# Q.3 - Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

### First of all, the mean has a greater range for the entire df than the subset. The department mean for the subset ranged from approximately 11 to 19, where the department mean for the entire df ranges from approximately 15 to 23.

### Second, the mean values for individual departments has changed as well. For example, the mean value for dept 14 was 19.46 in the subset, but is 16.77 in the entire df. This demonstrates that the calculating the mean value in a subset may be a helpful place to get a feel for the data, but is no substitute for an analysis of the ENTIRE DATA SET.

### Therefore, the only accurate mean value for each department are the values above, calculated from the entire dataframe.

# Q.4 - Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform( ) and loc( ) functions.

In [7]:
# Creating a max_order column for the entire df using transform

ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [8]:
# Checking for new column

ords_prods_merge.head(15)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329.0      1.0    prior           1.0                 2.0   
1   2398795.0      1.0    prior           2.0                 3.0   
2    473747.0      1.0    prior           3.0                 3.0   
3   2254736.0      1.0    prior           4.0                 4.0   
4    431534.0      1.0    prior           5.0                 4.0   
5   3367565.0      1.0    prior           6.0                 2.0   
6    550135.0      1.0    prior           7.0                 1.0   
7   3108588.0      1.0    prior           8.0                 1.0   
8   2295261.0      1.0    prior           9.0                 1.0   
9   2550362.0      1.0    prior          10.0                 4.0   
10  2968173.0     15.0    prior          15.0                 1.0   
11  1870022.0     15.0    prior          17.0                 2.0   
12  1911383.0     15.0    prior          18.0                 2.0   
13  2715276.0     15.0    prior          21.0                 1.0   
14   487368.0     15.0    prior          22.0                 1.0   

    order_time_of_day  days_since_prior_order first_time_ordering  product_id  \
0                 8.0                     NaN                True         196   
1                 7.0                    15.0               False         196   
2                12.0                    21.0               False         196   
3                 7.0                    29.0               False         196   
4                15.0                    28.0               False         196   
5                 7.0                    19.0               False         196   
6                 9.0                    20.0               False         196   
7                14.0                    14.0               False         196   
8                16.0                     0.0               False         196   
9                 8.0                    30.0               False         196   
10                9.0                     7.0               False         196   
11               16.0                     8.0               False         196   
12               11.0                     7.0               False         196   
13                9.0                     7.0               False         196   
14               10.0                    14.0               False         196   

    add_to_cart_order  ...  product_name aisle_id  department_id  prices  \
0                 1.0  ...          Soda     77.0            7.0     9.0   
1                 1.0  ...          Soda     77.0            7.0     9.0   
2                 1.0  ...          Soda     77.0            7.0     9.0   
3                 1.0  ...          Soda     77.0            7.0     9.0   
4                 1.0  ...          Soda     77.0            7.0     9.0   
5                 1.0  ...          Soda     77.0            7.0     9.0   
6                 1.0  ...          Soda     77.0            7.0     9.0   
7                 2.0  ...          Soda     77.0            7.0     9.0   
8                 4.0  ...          Soda     77.0            7.0     9.0   
9                 1.0  ...          Soda     77.0            7.0     9.0   
10                2.0  ...          Soda     77.0            7.0     9.0   
11                6.0  ...          Soda     77.0            7.0     9.0   
12                1.0  ...          Soda     77.0            7.0     9.0   
13                1.0  ...          Soda     77.0            7.0     9.0   
14                1.0  ...          Soda     77.0            7.0     9.0   

    _merge    price_range_loc     busiest_day    busiest_days  \
0     both  Mid-range product  Regularly busy  Regularly busy   
1     both  Mid-range product  Regularly busy    Slowest days   
2     both  Mid-range product  Regularly busy    Slowest days   
3     both  Mid-range product      Least busy    Slowest days   
4     both  Mid-range product      Least busy    Slowest days

In [9]:
# Creating loyalty flag using loc

ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [10]:
# Checking frequency of loyalty flag

ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15891507
Loyal customer      10294027
New customer         6249525
NaN                       11
Name: loyalty_flag, dtype: int64

# Q.5 - Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). 

In [11]:
# Aggregating for Mean, Min, Max for product prices using loyalty flag

ords_prods_merge.groupby('loyalty_flag').agg({'prices':['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

# Q.6 - Create a spending flag for each user based on the average price across all their orders using the following criteria:

## A: If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”

## B: If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [12]:
# First, creating new column for spending, using mean

ords_prods_merge['spending'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [13]:
ords_prods_merge.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0  2539329.0      1.0    prior           1.0                 2.0   
1  2398795.0      1.0    prior           2.0                 3.0   
2   473747.0      1.0    prior           3.0                 3.0   
3  2254736.0      1.0    prior           4.0                 4.0   
4   431534.0      1.0    prior           5.0                 4.0   

   order_time_of_day  days_since_prior_order first_time_ordering  product_id  \
0                8.0                     NaN                True         196   
1                7.0                    15.0               False         196   
2               12.0                    21.0               False         196   
3                7.0                    29.0               False         196   
4               15.0                    28.0               False         196   

   add_to_cart_order  ...  department_id prices  _merge    price_range_loc  \
0                1.0  ...            7.0    9.0    both  Mid-range product   
1                1.0  ...            7.0    9.0    both  Mid-range product   
2                1.0  ...            7.0    9.0    both  Mid-range product   
3                1.0  ...            7.0    9.0    both  Mid-range product   
4                1.0  ...            7.0    9.0    both  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day max_order  \
0  Regularly busy  Regularly busy        Average orders      10.0   
1  Regularly busy    Slowest days        Average orders      10.0   
2  Regularly busy    Slowest days           Most orders      10.0   
3      Least busy    Slowest days        Average orders      10.0   
4      Least busy    Slowest days           Most orders      10.0   

   loyalty_flag  spending  
0  New customer  6.367797  
1  New customer  6.367797  
2  New customer  6.367797  
3  New customer  6.367797  
4  New customer  6.367797  

[5 rows x 23 columns]

In [14]:
# Second, using loc function to create flag for spending habits

ords_prods_merge.loc[ords_prods_merge['spending'] < 10, 'type_of_spender'] = 'Low spender'
ords_prods_merge.loc[ords_prods_merge['spending'] >= 10, 'type_of_spender'] = 'High spender'

In [15]:
ords_prods_merge.head(15)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329.0      1.0    prior           1.0                 2.0   
1   2398795.0      1.0    prior           2.0                 3.0   
2    473747.0      1.0    prior           3.0                 3.0   
3   2254736.0      1.0    prior           4.0                 4.0   
4    431534.0      1.0    prior           5.0                 4.0   
5   3367565.0      1.0    prior           6.0                 2.0   
6    550135.0      1.0    prior           7.0                 1.0   
7   3108588.0      1.0    prior           8.0                 1.0   
8   2295261.0      1.0    prior           9.0                 1.0   
9   2550362.0      1.0    prior          10.0                 4.0   
10  2968173.0     15.0    prior          15.0                 1.0   
11  1870022.0     15.0    prior          17.0                 2.0   
12  1911383.0     15.0    prior          18.0                 2.0   
13  2715276.0     15.0    prior          21.0                 1.0   
14   487368.0     15.0    prior          22.0                 1.0   

    order_time_of_day  days_since_prior_order first_time_ordering  product_id  \
0                 8.0                     NaN                True         196   
1                 7.0                    15.0               False         196   
2                12.0                    21.0               False         196   
3                 7.0                    29.0               False         196   
4                15.0                    28.0               False         196   
5                 7.0                    19.0               False         196   
6                 9.0                    20.0               False         196   
7                14.0                    14.0               False         196   
8                16.0                     0.0               False         196   
9                 8.0                    30.0               False         196   
10                9.0                     7.0               False         196   
11               16.0                     8.0               False         196   
12               11.0                     7.0               False         196   
13                9.0                     7.0               False         196   
14               10.0                    14.0               False         196   

    add_to_cart_order  ...  prices _merge    price_range_loc     busiest_day  \
0                 1.0  ...     9.0   both  Mid-range product  Regularly busy   
1                 1.0  ...     9.0   both  Mid-range product  Regularly busy   
2                 1.0  ...     9.0   both  Mid-range product  Regularly busy   
3                 1.0  ...     9.0   both  Mid-range product      Least busy   
4                 1.0  ...     9.0   both  Mid-range product      Least busy   
5                 1.0  ...     9.0   both  Mid-range product  Regularly busy   
6                 1.0  ...     9.0   both  Mid-range product  Regularly busy   
7                 2.0  ...     9.0   both  Mid-range product  Regularly busy   
8                 4.0  ...     9.0   both  Mid-range product  Regularly busy   
9                 1.0  ...     9.0   both  Mid-range product      Least busy   
10                2.0  ...     9.0   both  Mid-range product  Regularly busy   
11                6.0  ...     9.0   both  Mid-range product  Regularly busy   
12                1.0  ...     9.0   both  Mid-range product  Regularly busy   
13                1.0  ...     9.0   both  Mid-range product  Regularly busy   
14                1.0  ...     9.0   both  Mid-range product  Regularly busy   

      busiest_days busiest_period_of_day max_order      loyalty_flag  \
0   Regularly busy        Average orders      10.0      New customer   
1     Slowest days        Average orders      10.0      New customer   
2     Slowest days           Most orders      10.0      New customer   
3     Slowest days        Average 

In [16]:
# Using a frequency check for type_of_spender

ords_prods_merge['type_of_spender'].value_counts(dropna = False)

Low spender     31800509
High spender      634550
NaN                   11
Name: type_of_spender, dtype: int64

# Q.7 - Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. Use the following criteria:

## A: If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”

## B: If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”

## C: If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [17]:
# First, creating new column for order_freq, using median

ords_prods_merge['order_freq'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [18]:
ords_prods_merge.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0  2539329.0      1.0    prior           1.0                 2.0   
1  2398795.0      1.0    prior           2.0                 3.0   
2   473747.0      1.0    prior           3.0                 3.0   
3  2254736.0      1.0    prior           4.0                 4.0   
4   431534.0      1.0    prior           5.0                 4.0   

   order_time_of_day  days_since_prior_order first_time_ordering  product_id  \
0                8.0                     NaN                True         196   
1                7.0                    15.0               False         196   
2               12.0                    21.0               False         196   
3                7.0                    29.0               False         196   
4               15.0                    28.0               False         196   

   add_to_cart_order  ...  _merge    price_range_loc     busiest_day  \
0                1.0  ...    both  Mid-range product  Regularly busy   
1                1.0  ...    both  Mid-range product  Regularly busy   
2                1.0  ...    both  Mid-range product  Regularly busy   
3                1.0  ...    both  Mid-range product      Least busy   
4                1.0  ...    both  Mid-range product      Least busy   

     busiest_days  busiest_period_of_day max_order  loyalty_flag  spending  \
0  Regularly busy         Average orders      10.0  New customer  6.367797   
1    Slowest days         Average orders      10.0  New customer  6.367797   
2    Slowest days            Most orders      10.0  New customer  6.367797   
3    Slowest days         Average orders      10.0  New customer  6.367797   
4    Slowest days            Most orders      10.0  New customer  6.367797   

  type_of_spender order_freq  
0     Low spender       20.5  
1     Low spender       20.5  
2     Low spender       20.5  
3     Low spender       20.5  
4     Low spender       20.5  

[5 rows x 25 columns]

In [19]:
# Creating frequency flag using loc function

ords_prods_merge.loc[ords_prods_merge['order_freq'] > 20, 'frequency_flag'] = 'Non-frequent customer'
ords_prods_merge.loc[(ords_prods_merge['order_freq'] > 10) & (ords_prods_merge['order_freq'] <= 20), 'frequency_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['order_freq'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [21]:
ords_prods_merge.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0  2539329.0      1.0    prior           1.0                 2.0   
1  2398795.0      1.0    prior           2.0                 3.0   
2   473747.0      1.0    prior           3.0                 3.0   
3  2254736.0      1.0    prior           4.0                 4.0   
4   431534.0      1.0    prior           5.0                 4.0   

   order_time_of_day  days_since_prior_order first_time_ordering  product_id  \
0                8.0                     NaN                True         196   
1                7.0                    15.0               False         196   
2               12.0                    21.0               False         196   
3                7.0                    29.0               False         196   
4               15.0                    28.0               False         196   

   add_to_cart_order  ...    price_range_loc     busiest_day    busiest_days  \
0                1.0  ...  Mid-range product  Regularly busy  Regularly busy   
1                1.0  ...  Mid-range product  Regularly busy    Slowest days   
2                1.0  ...  Mid-range product  Regularly busy    Slowest days   
3                1.0  ...  Mid-range product      Least busy    Slowest days   
4                1.0  ...  Mid-range product      Least busy    Slowest days   

   busiest_period_of_day  max_order  loyalty_flag  spending type_of_spender  \
0         Average orders       10.0  New customer  6.367797     Low spender   
1         Average orders       10.0  New customer  6.367797     Low spender   
2            Most orders       10.0  New customer  6.367797     Low spender   
3         Average orders       10.0  New customer  6.367797     Low spender   
4            Most orders       10.0  New customer  6.367797     Low spender   

  order_freq         frequency_flag  
0       20.5  Non-frequent customer  
1       20.5  Non-frequent customer  
2       20.5  Non-frequent customer  
3       20.5  Non-frequent customer  
4       20.5  Non-frequent customer  

[5 rows x 26 columns]

In [22]:
ords_prods_merge['frequency_flag'].value_counts(dropna = False)

Frequent customer        21578273
Regular customer          7217066
Non-frequent customer     3639720
NaN                            11
Name: frequency_flag, dtype: int64

# Q.9 - Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder.

In [23]:
# Exporting ords_prods_merge df as "opm_aggr"

ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'opm_aggr.pkl'))